plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/PassthroughQueryPageSource.java

```diff
@@ -87,5 +114,121 @@ public long getSystemMemoryUsage()
     public void close()
             throws IOException
     {
+        iterator.close();
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    private static class ScrollSearchResponse
+    {
+        @JsonProperty("_scroll_id")
+        private String scrollId;
+
+        @JsonProperty("hits")
+        private Hits hits;
+
+        public String getScrollId()
+        {
+            return scrollId;
+        }
+
+        public boolean hasMoreHits()
+        {
+            return hits != null && hits.getHits() != null && !hits.getHits().isEmpty();
+        }
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    private static class Hits
+    {
+        @JsonProperty("hits")
+        private List<Object> hits;
+
+        public List<Object> getHits()
+        {
+            return hits;
+        }
+    }
+
+    private static class SearchHitIterator
+            extends AbstractIterator<String>
+    {
+        private final ElasticsearchClient client;
+        private final String first;
+
+        private String scrollId;
+        private Boolean hasMoreHits;
+        private long readTimeNanos;
+        String response;
+        ObjectMapper objectMapper;
+
+        public SearchHitIterator(ElasticsearchClient client, String first)
+        {
+            this.client = client;
+            this.first = first;
+            this.objectMapper = new ObjectMapper();
+        }
+
+        public long getReadTimeNanos()
+        {
+            return readTimeNanos;
+        }
+
+        @Override
+        protected String computeNext()
+        {
+            long start = System.nanoTime();
+
+            if (scrollId == null) {
+                response = first;
+            }
+            else {
+                response = client.executeScrollQuery(scrollId);
+            }
+
+            readTimeNanos += System.nanoTime() - start;
+            reset(response);
+
+            if (!hasMoreHits || scrollId == null) {
+                return endOfData();
+            }
+
+            return response;
+        }
+
+        private void reset(String response)
+        {
+            ScrollSearchResponse scrollSearchResponse = extractSearchResponse(response);
+            if (scrollSearchResponse != null) {
+                scrollId = scrollSearchResponse.getScrollId();
+                hasMoreHits = scrollSearchResponse.hasMoreHits();
+            }
+        }
+
+        private ScrollSearchResponse extractSearchResponse(String response)
+        {
+            try {
+                ScrollSearchResponse scrollSearchResponse = objectMapper.readValue(response, ScrollSearchResponse.class);
```

Comment:
you can use a static final `JsonCodec<ScrollSearchResponse>`  object  like other places in the code